test/modeling_library/distributions.jl

```diff
@@ -126,9 +126,9 @@ end
     f = (x, mu, std) -> logpdf(broadcasted_normal, x, mu, std)
     args = (fill(0.4), fill(0.2), fill(0.3))
     actual = logpdf_grad(broadcasted_normal, args...)
-    @test isapprox(actual[1], finite_diff(f, args, 1, dx; broadcast=true))
-    @test isapprox(actual[2], finite_diff(f, args, 2, dx; broadcast=true))
-    @test isapprox(actual[3], finite_diff(f, args, 3, dx; broadcast=true))
```

Comment:
Why this change?

Reply:
I'm actually a little unsure how to fix these tests. It seems that the first argument of `isapprox` is a zero-dimensional array (which I believe is the expected behavior) but the second argument is a `Float64`, which causes Julia to complain that could no method of `isapprox` mathes the given argument types. Any suggestions on how to fix this?

Reply:
After investigating this a bit, I think it boils down to https://github.com/JuliaLang/julia/issues/28866, which I consider to be a bug in Julia.  Yes -- in the version of the code before this PR, both the LHS and the RHS should be 0-dimensional arrays.  For the RHS, the issue appears to be that on these lines
https://github.com/probcomp/Gen.jl/blob/d79aded6909a8e44b3a813179758add8237ed46a/test/runtests.jl#L22-L23
do `::Aray{T,0} .+ ::T` and `::Aray{T,0} .- ::T`, which should give an `::Array{T,0}` but due to https://github.com/JuliaLang/julia/issues/28866 gives a scalar.  So the RHS of the `./` on the following line should also be an `Array{T,0}`, and the output therefore an `Array{T,0}`, but it's not and it's not.

I think the cleanest way to fix this is to add a workaround to the (test-only) function `finite_diff` so that it handles zero-dimensional array valued arguments correctly.

Reply:
Oh whoops, I take that back -- both sides of the `./` are scalars:
https://github.com/probcomp/Gen.jl/blob/d79aded6909a8e44b3a813179758add8237ed46a/test/runtests.jl#L24
I guess what my previous comment really shows is that, if `f` requires its argument to be a zero-dimensional array and errors if given a scalar, `finite_diff` will hit that error.

A fix to `finite_diff` for the zero-dimensional case still seems right :slightly_smiling_face:

Reply:
(I'm currently drafting this.)

Reply:
Aha, in addition to the above, the function `f` being supplied to `finite_diff` does not satisfy the requirements in its docstring: it takes array-valued arguments, but is not the broadcast of a function that takes scalar-valued arguments (i.e. `logpdf_grad(vector-valued distribution parameters)` is not a broadcast of a bunch of `logpdf_grad(scalar-valued distribution parameters)`'s).  So I guess we should add a variant of `finite_diff` that has the semantics we need.

Reply:
Yeah, I think to properly test this, we will need a more general `finite_diff` function or use an autodiff library implementation as a reference.

Reply:
you clearly trust autodiff more than I do :slightly_smiling_face: